transaction/core/src/membership_proofs/mod.rs

```diff
@@ -50,6 +50,141 @@ fn hash_nil() -> [u8; 32] {
+
+    result.hash.0 = hash_nodes(&left, &right);
+    Ok(result)
+}
+
+// Helper: Conditionally assign to [u8; 32]
+//
+// It seems subtle doesn't implement ConditionallySelectable on [u8; 32], not sure why,
+// so we can't just use the conditional_assign function from that API.
+//
+// If this is slow then we can later used a faster implementation, but it probably won't be that slow.
+fn conditional_assign_32_bytes(target: &mut [u8; 32], src: &[u8; 32], cond: Choice) {
+    for idx in 0..32 {
+        target[idx].conditional_assign(&src[idx], cond);
+    }
+}
+
+/// Checks that a proof of membership is well-formed, and returns the hash that it implies for the root.
+/// This hash should then be checked against known root hash value.
+///
+/// Errors if:
+/// - Any proof elements have invalid ranges (from and to out of order)
+/// - The first element is missing or its range doesn't match proof.index
+/// - Any of the proof elements could not be combined with result of combining predecessors
+///
+/// Does not read the proof.highest_index value or compare it with the range
+/// that we compute by combining elements. But this should not be strictly necessary
+/// for secure merkle proof validation.
+///
+/// Note: This is pub in order to allow that it can be used in debugging assertions elsewhere.
+/// This could simply be a member function on TxOutMembershipProof, but that would require
+/// hash_nodes function to be in scope in that module, so for now we didn't do that.
+pub fn compute_implied_merkle_root(
+    proof: &TxOutMembershipProof,
+) -> Result<TxOutMembershipElement, Error> {
+    // All Ranges contained in the proof must be valid. An invalid Range could be created
+    // by deserializing invalid bytes.
+    if proof.elements.iter().any(|e| e.range.from > e.range.to) {
+        return Err(Error::RangeError(RangeError {}));
+    }
+
+    // The first element should correspond to proof.index in a well-formed proof
+    let first = proof
+        .elements
+        .first()
+        .ok_or_else(|| Error::MissingLeafHash(proof.index))?
+        .clone();
+    if first.range.from != proof.index || first.range.to != proof.index {
+        return Err(Error::MissingLeafHash(proof.index));
+    }
+
+    // Try to fold subsequent elements together with the first element,
+    // combining them using `combine_adjacent_membership_elements`.
+    // idx is used only to provide context for the error message
+    let implied_root = proof.elements[1..].iter().enumerate().try_fold(
+        first,
+        |prev, (idx, next)| -> Result<TxOutMembershipElement, Error> {
+            compose_adjacent_membership_elements(&prev, next)
+                .map_err(|_| Error::UnexpectedMembershipElement(idx + 1))
+        },
+    )?;
+
+    // At this point, we could check that implied_root.range.to = proof.highest_index, or that -1, or something,
+    // but I'm not sure at this moment what exactly the right test is, and I don't think we really need this,
```

Comment:
What would you need to be able to be sure what the right test is? Or how could we resolve the uncertainty around whether we need to test any other qualities?